crates/ra_ide_api/src/completion/complete_path.rs

```diff
@@ -10,7 +10,7 @@ pub(super) fn complete_path(acc: &mut Completions, ctx: &CompletionContext) {
         Some(path) => path.clone(),
         _ => return,
     };
-    let def = match ctx.resolver.resolve_path(ctx.db, &path).take_types() {
+    let def = match ctx.resolver.resolve_path(ctx.db, &path).into_per_ns().take_types() {
```

Comment:
I think it's important that nothing outside the HIR observes "partial" results. Let's make `resolve_path` a `pub` method that returns either a complete resolution or a failure, and instead introduce a second `pub(crate) fn resolve_path_segments` or something that resolves paths partially. 

As usual, we should be careful when using types (like Resolver in this case) both as a public API and internally in the implementation, lest the impl detail leak to the `pub` API. `pub(crate)` is totally OK though!

Reply:
What kind of failure should `resolve_path` return ? Just `PerNs::none` or something more specific?

Reply:
I think `PerNs::none` is definitely a good start! We can always add someting more specific later.

Reply:
Should be fixed in 247d1c1. 

Now `resolve_path` will only return the fully resolved path or `PerNs::none`, made `resolve_path_segments` and `PathResult` both `pub(crate)` as they should be internal to the crate.